### PR TITLE
TRD gain calibration

### PR DESCRIPTION
**Pull request by glegras**

### CODE REVIEW

File: Detectors/TRD/calibration/include/TRDCalibration/CalibratorGain.h
Comment (f3sch) [inline Detectors/TRD/calibration/include/TRDCalibration/CalibratorGain.h:71]: You are not doing that. Why define at all, just initialize with your default value.

File: Detectors/TRD/calibration/src/CalibratorGain.cxx
Comment (f3sch) [inline Detectors/TRD/calibration/src/CalibratorGain.cxx]: This seems somewhat wasteful. You are never freeing the memory allocated to these functions and continuously allocating more. How about moving the  fit function to class member, while performing the convolution once during the init?

File: Detectors/TRD/calibration/src/CalibratorGain.cxx
Comment (f3sch) [inline Detectors/TRD/calibration/src/CalibratorGain.cxx]: Btw, how stable are your fits? I struggled somewhat with this and achieved stability with setting parameter limits like these here:
``` C
     fitFunction->SetParLimits(0, 30., 100.0);
      fitFunction->SetParLimits(1, 5.0, 25.0);
      fitFunction->SetParLimits(2, -0.1, 0.5);
      fitFunction->SetParLimits(3, -0.1, 0.5);
      fitFunction->SetParLimits(5, 1.0, 10.0);
```

File: Detectors/TRD/calibration/src/CalibratorGain.cxx
Comment (f3sch) [inline Detectors/TRD/calibration/src/CalibratorGain.cxx]: If you use parameter limits, I think you must add "B" to the options here.

File: Detectors/TRD/calibration/src/CalibratorGain.cxx
Comment (f3sch) [inline Detectors/TRD/calibration/src/CalibratorGain.cxx]:
```suggestion
      LOGF(warn, "Fit for chamber %i failed, nEntries: %d", iDet, nEntries);
```

File: Detectors/TRD/workflow/include/TRDWorkflow/GainCalibSpec.h
Comment (f3sch) [inline Detectors/TRD/workflow/include/TRDWorkflow/GainCalibSpec.h]: This mechanism has changed upstream since you last looked, for reference see #11155.

File: Detectors/TRD/workflow/io/src/TRDCalibWriterSpec.cxx
Comment (f3sch) [inline Detectors/TRD/workflow/io/src/TRDCalibWriterSpec.cxx:14]: @martenole Doe this warrant its own writer/reader can we not merge the for the vdrift calib into one, i.e., having one file like trdcalib.root?

File: Detectors/TRD/calibration/src/TrackBasedCalib.cxx
Comment (f3sch) [inline Detectors/TRD/calibration/src/TrackBasedCalib.cxx]: @pachmay, @tdietel  should the gain calibration use the krypton calibration, it seemed to me that they were independent?

File: Detectors/TRD/calibration/src/TrackBasedCalib.cxx
Comment (f3sch) [inline Detectors/TRD/calibration/src/TrackBasedCalib.cxx:88]: How about moving this to a class member and removing all the calls to Instance()?

File: Detectors/TRD/calibration/src/TrackBasedCalib.cxx
Comment (f3sch) [inline Detectors/TRD/calibration/src/TrackBasedCalib.cxx:141]: I think you should move this check before applying the corrections.

**Review by f3sch — DISMISSED** [review]
Nice work :+1: ! I added some general comments, questions and I am sure others will review much more thoroughly.

**Comment by martenole** [comment]
Hi, was there a presentation with the details on the gain calibration somewhere? Could you point me to that or elaborate here a bit more what you plan to do? In general its of course very good to see progress here, so thank you for your work.

File: Detectors/TRD/calibration/src/TrackBasedCalib.cxx
Comment (martenole) [reply Detectors/TRD/calibration/src/TrackBasedCalib.cxx:141]: why do you actually want to ignore such values here?

File: Detectors/TRD/calibration/src/TrackBasedCalib.cxx
Comment (martenole) [inline Detectors/TRD/calibration/src/TrackBasedCalib.cxx]: I don't think a full propagation is necessary here. This would duplicate a lot of work done anyhow for the vDrift and ExB calibration so it should either be combined or you just use a simple `getSnpAt(radius, b-field)` to obtain the track inclination at given radius. The eta should be constant enough in any case so you can get that from the outer param of the TRD track directly

File: Detectors/TRD/calibration/src/TrackBasedCalib.cxx
Comment (martenole) [inline Detectors/TRD/calibration/src/TrackBasedCalib.cxx]: the trigonometric functions and especially the `asin` operation are rather slow, so you should check for an alternative in places which are called often like here for every track. Instead of `tan (phi)^2` you can substitute `sin (phi)^2 / (1 - sin(phi)^2)` so you can use the parameter stored in the track directly

File: Detectors/TRD/workflow/io/src/TRDCalibWriterSpec.cxx
Comment (martenole) [reply Detectors/TRD/workflow/io/src/TRDCalibWriterSpec.cxx:14]: I am not sure if we need reader/writer at all?

**Review by martenole — DISMISSED** [review]
Some comments already now and the (**dummy!**) approval to get the CI checks started for first time contributors. I think we should discuss this, for example  at the next TRD meeting in a bit more detail

**Comment by glegras** [comment]
Hi, I think that indeed, it is better to discuss it at next TRD meeting in more detail.

Let me also remind the context: We want to have automatized what I did locally (among others this presentation at this TRD meeting in January: https://indico.cern.ch/event/1248149/contributions/5244607/attachments/2584311/4457974/TRDGainCalibration.pdf )

And a description (duplicate of the mail I sent to some of you):  
For the code, I just started from the code on drift velocity calibration as a base. During global tracking within the enable-trackbased-calib option, I collect for each tracklet (after cut on ntracklets, p, and dEdxTPC of the track), the dedx in the chamber (which is just sum of the charges, corrected by local gain and track length in the chamber). The output is for each timestamp an array with number of counts in each dedx bin and for each chamber.
Then, in the calibration workflow, the fit is done (with Landau*exp convoluted with Gaussian) for every chamber and we write the Most Probable Value.

File: Detectors/TRD/workflow/io/src/TRDCalibWriterSpec.cxx
Comment (pachmay) [reply Detectors/TRD/workflow/io/src/TRDCalibWriterSpec.cxx:14]: Hi @martenole, I am wondering: the gain calibration should run on the tracklet charges. If we restrict it to the digits, don't we then reduce the statistics a lot. When running on tracklets in the synchronous reco, we could obtain more statistics; however we need to merge the split charges before.

File: Detectors/TRD/calibration/src/TrackBasedCalib.cxx
Comment (pachmay) [reply Detectors/TRD/calibration/src/TrackBasedCalib.cxx]: the final tracklet charges used for PID should be corrected for chamber wise gain and local gain (krypton)

we could also correct the charges with the local gain before the chamber wise gain calibration, but I assume that there is not much difference

File: Detectors/TRD/workflow/io/src/TRDCalibWriterSpec.cxx
Comment (martenole) [reply Detectors/TRD/workflow/io/src/TRDCalibWriterSpec.cxx:14]: I think in this implementation the digits are not used at all and the tracklet charges are used instead. This is how it was planned, right?

Comment (pachmay) [reply Detectors/TRD/workflow/io/src/TRDCalibWriterSpec.cxx:14]: Maybe I misunderstood your earlier comment. Indeed the tracklet charges should be used; thus to get more statistics, it might be better to not restrict ourselves to the digit events and therefore we do not need/use the writer and reader for the drift velocity; is that technically possible?

Comment (pachmay) [reply Detectors/TRD/workflow/io/src/TRDCalibWriterSpec.cxx:14]: Hi Ole, maybe we could just do something similar as for the t0 calibration; we wait until enough statistics is accumulated and the fits are very successful and then proceed with the gain calibration. Would that work?

**Comment by f3sch** [comment]
@glegras could you make this for now a draft, then we would now when you want to review and merge this?

**Comment by glegras** [comment]
Hi, this is kind of my final version with your comments implemented. So you can start to review now.

File: Detectors/TRD/workflow/include/TRDWorkflow/GainCalibSpec.h
Comment (glegras) [inline Detectors/TRD/workflow/include/TRDWorkflow/GainCalibSpec.h]: Here, the line is supposed to be used for retrieving the latest results from the ccdb (used by the function retrievePrev in CalibratorGain.cxx). However, since the ccdb entry is not yet created, it returns an error for now. So I assume it should be left commented for now, and get uncommented later. (the same goes for the function retrievePrev)

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/GainCalibHistos.h
Comment (martenole) [inline DataFormats/Detectors/TRD/include/DataFormatsTRD/GainCalibHistos.h]: if you don't need this function, could you remove it?

File: DataFormats/Detectors/TRD/src/GainCalibHistos.cxx
Comment (martenole) [inline DataFormats/Detectors/TRD/src/GainCalibHistos.cxx]: this function will always return zero. What happens in case iBin is out of bounds?

File: Detectors/TRD/calibration/include/TRDCalibration/CalibratorGain.h
Comment (martenole) [inline Detectors/TRD/calibration/include/TRDCalibration/CalibratorGain.h]:
```suggestion
/// \author Gauthier Legras
```

File: Detectors/TRD/calibration/src/CalibratorGain.cxx
Comment (martenole) [inline Detectors/TRD/calibration/src/CalibratorGain.cxx]: We can upload a default object of the gain calibration before we merge your PR. So you can uncomment these lines already now.

File: Detectors/TRD/calibration/src/CalibratorGain.cxx
Comment (martenole) [inline Detectors/TRD/calibration/src/CalibratorGain.cxx]: all class member variables must start with `m`, so `mFconv` for example would be valid. Also, I can you replace the plain `new` and use smart pointers instead?

File: Detectors/TRD/calibration/src/CalibratorGain.cxx
Comment (martenole) [inline Detectors/TRD/calibration/src/CalibratorGain.cxx]: same as above

File: Detectors/TRD/calibration/src/CalibratorGain.cxx
Comment (martenole) [reply Detectors/TRD/calibration/src/CalibratorGain.cxx]: should be OK without as long as no predefined function is used for fitting: https://root.cern.ch/doc/master/classTH1.html#a7e7d34c91d5ebab4fc9bba3ca47dabdd

File: Detectors/TRD/calibration/src/CalibratorGain.cxx
Comment (martenole) [inline Detectors/TRD/calibration/src/CalibratorGain.cxx]: there is no need for the extra protection and -10 seconds

File: Detectors/TRD/calibration/src/CalibratorGain.cxx
Comment (martenole) [inline Detectors/TRD/calibration/src/CalibratorGain.cxx]: I would set the end validity to + 3 days. By default, the objects will be uploaded to the CCDB with adjustable end of validity. This means that the object you upload will in the beginning have a long validity, but the moment a new object is uploaded the end validity of the previous object is set to the start validity of the new one. So the validities don't overlap and you avoid gaps. I should do the same for the vD and ExB calibration

File: Detectors/TRD/calibration/src/CalibratorGain.cxx
Comment (martenole) [reply Detectors/TRD/calibration/src/CalibratorGain.cxx]: Actually, what do you plan to do with the previous values which you try to retrieve here?

File: Detectors/TRD/workflow/include/TRDWorkflow/GainCalibSpec.h
Comment (martenole) [reply Detectors/TRD/workflow/include/TRDWorkflow/GainCalibSpec.h]: we can upload a default object already now

File: Detectors/TRD/workflow/include/TRDWorkflow/GainCalibSpec.h
Comment (martenole) [inline Detectors/TRD/workflow/include/TRDWorkflow/GainCalibSpec.h]: the comment does not make sense here

File: Detectors/TRD/workflow/include/TRDWorkflow/GainCalibSpec.h
Comment (martenole) [inline Detectors/TRD/workflow/include/TRDWorkflow/GainCalibSpec.h]: seems you are not using these aliases, could you please remove them?

File: Detectors/TRD/workflow/src/TrackBasedCalibSpec.cxx
Comment (martenole) [inline Detectors/TRD/workflow/src/TrackBasedCalibSpec.cxx]: You will need to add some logic here. So far the track based calibration was only important for the vDrift and ExB calibration. Now you are adding gain, but we don't have both calibrations enabled always. So the track based calibration needs to be aware which calibration should be run and input/output and processing should be configured accordingly

File: Detectors/TRD/calibration/src/TrackBasedCalib.cxx
Comment (martenole) [inline Detectors/TRD/calibration/src/TrackBasedCalib.cxx:151]: can you please add a comment here how you calculate the track length and double check that this is correct?

File: Detectors/TRD/calibration/src/TrackBasedCalib.cxx
Comment (martenole) [inline Detectors/TRD/calibration/src/TrackBasedCalib.cxx]:
```suggestion
        localGainCorr = mLocalGain->getValue(tracklet.getDetector(), tracklet.getPadCol(), tracklet.getPadRow());
```

**Review by martenole — CHANGES_REQUESTED** [review]
Thanks, I think we are on a good track here. The only larger task which is still missing is to add some logic to the TrackBasedCalib class. Let me know in case some comments are not clear

File: Detectors/TRD/calibration/src/CalibratorGain.cxx
Comment (glegras) [reply Detectors/TRD/calibration/src/CalibratorGain.cxx]: I was thinking that uploading the previous value when not enough statistics should be better that uploading the default value.
But in fact, we could also decide to do an interpolation, so maybe it is better to let it commented (or even remove it and let the default value) and discuss it later.

File: Detectors/TRD/calibration/src/CalibratorGain.cxx
Comment (glegras) [reply Detectors/TRD/calibration/src/CalibratorGain.cxx]: What do you exactly want me to do with smart pointers?

File: Detectors/TRD/workflow/src/TrackBasedCalibSpec.cxx
Comment (glegras) [reply Detectors/TRD/workflow/src/TrackBasedCalibSpec.cxx]: I would think by by default we will always doing both calibrations during reconstruction?
But okay, I can still add the options enable-vdexb-calib and enable-gain-calib
In this case, should I let the option enable-trackbased-calib for post compatibility, or can I remove it?

File: Detectors/TRD/calibration/src/CalibratorGain.cxx
Comment (martenole) [reply Detectors/TRD/calibration/src/CalibratorGain.cxx]: you should avoid a plain "new" in your code. Here for example you are never deleting fConv and fitFunction. Even though both are created in the init function which is called only once this is bad practice and should be avoided

File: Detectors/TRD/calibration/src/CalibratorGain.cxx
Comment (martenole) [reply Detectors/TRD/calibration/src/CalibratorGain.cxx]: I would not complicate things with an interpolation. Taking the value from the previous calibration is OK. As said, we can upload a default value beforehand and then you can uncomment this block already now

File: Detectors/TRD/workflow/src/TrackBasedCalibSpec.cxx
Comment (martenole) [reply Detectors/TRD/workflow/src/TrackBasedCalibSpec.cxx]: By default we definitely want the vDrift and ExB calibration during the synchronous reco, since they have an impact on the tracklets which are used for global tracking and the TPC calibration. During the asynchronous reco we would want both to be active.
The enable-trackbased-calib option should stay. And you should add another option to the trd-tracking-workflow named enable-gain-calib for example. This boolean flag (false by default) should be passed to the track based calib device which then either adds the gain calibration or not. 
So in case trackbased calibrations are enabled we will always have vD and ExB, but gain will be optional

File: Detectors/TRD/calibration/src/CalibratorGain.cxx
Comment (glegras) [reply Detectors/TRD/calibration/src/CalibratorGain.cxx]: Hi @martenole, in this case can you upload the object already now or tell me how to do it, so that I can uncomment and test the code before updating the PR?

Comment (martenole) [reply Detectors/TRD/calibration/src/CalibratorGain.cxx]: Sure. To what should I set the default? 42?

Comment (martenole) [reply Detectors/TRD/calibration/src/CalibratorGain.cxx]: Hi @glegras can you check if this is OK? http://ccdb-test.cern.ch:8080/browse/TRD/Calib/CalibGain

Comment (glegras) [reply Detectors/TRD/calibration/src/CalibratorGain.cxx]: Hi @martenole, can you change the start validity date of the default object to beginning of 2022?

**Review by martenole — DISMISSED** [review]
DON'T MERGE YET, approving so that the CI starts

File: Detectors/TRD/workflow/src/TrackBasedCalibSpec.cxx
Comment (martenole) [inline Detectors/TRD/workflow/src/TrackBasedCalibSpec.cxx]: The changes you do here to the DPL device should depend on whether the gain calibration was enabled or not. If not, then no additional inputs/outputs should be requested

**Review by martenole — COMMENTED** [review]
Hi Gauthier, this looks very good and we can almost merge it as is. You will only need to change the `Detectors/TRD/workflow/src/trd-tracking-workflow.cxx` and add the option `enable-gain-calib` there instead of to the TrackBasedCalib.
Because from that workflow you can pass the option down to the TrackBasedCalib device and then decide there which input/output should be requested.
Then I think we are good to go

**Review by martenole — APPROVED** [review]
Approving so that the CI checks are started.
@glegras I had to add some changes on top, especially because the large `std::array<int, nChambers*nBins>` was not feasible in the `GainCalibHistos` class. Now it should be OK. I did a small test reconstruction just to see that the workflow does not crash. But I did not check yet whether the output is meaningful.
In any case, when the CI is green we can merge it as is and then you can continue to commission the gain calibration.
Cheers,
Ole

**Pull request merged**
